### PR TITLE
Add related and visited NFTs

### LINKED ISSUE #990: Related strip of NFTs contextual to NFT item
Continue from 
- https://github.com/kodadot/nft-gallery/issues/782
- https://github.com/kodadot/nft-gallery/issues/478

Make component as two rows, both scrollable

First would show more from the current collection with a horizontal scroll, limit to 3-10 NFTs, show only BUY now and recent minted, newest and cheapest?

Second would show previously viewed NFTs, which would be stored at users localstorage, with 3-10 last NFTs viewed

### PR DESCRIPTION
**Pull request by preschian**
**Thank you for your contribution** to the [KodaDot NFT gallery](https://kodadot.xyz).
👇 \_ Let's make a quick check before the contribution.

### PR type

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring

### What's new?

- [x] PR closes #990 

### Before submitting Pull Request, please make sure:

- [x] My contribution builds **clean without any errors or warnings**
- [x] I've merged the recent default branch -- **main** and I've no conflicts
- [x] I've tried to respect high code quality standards
- [x] I've didn't break any original functionality
- [x] I've posted a screenshot of demonstrated change in this PR

### Optional

- [ ] I've tested it at </rmrk/collection/26902bc2f7c20c546a-1FVG7>
- [ ] I've tested PR on mobile and everything seems works
- [ ] I found edge cases
- [ ] I've written some unit tests 🧪

### Had issue bounty label?

- [x] Fill up your KSM address: [Payout](https://beta.kodadot.xyz/transfer/?target=DY4SQF2iD456tH89aQtz5wv1EV3BbSW8wKKuMcwbmXaj1pM)

### Community participation

- [x] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

### Screenshot

- [x] My fix has changed **something** on UI; a screenshot is best to understand changes for others.

Screen Shot 2022-03-31 at 9 17 29 PM

### CODE REVIEW

**Comment by yangwao** [comment]
oh nice! 

https://deploy-preview-2722--koda-nuxt.netlify.app/rmrk/gallery/11275828-800f8a914281765a7d-KITTY-KITTY_PARADISE_262-0000000000000262

**Comment by yangwao** [comment]
@preschian are you showing in the first row the sorted by cheapest ? 

image

https://deploy-preview-2722--koda-nuxt.netlify.app/rmrk/gallery/10979592-8c5359d2bddaca9f4c-BBSF-COMMAND_CHIEF_MASTER_SERGEANT_FROMAGE-0000000000000080

**Comment by preschian** [comment]
Updated with sorted by cheapest, and then newest. `orderBy: [price_ASC, updatedAt_DESC]`. Is the graphql query correct?

Screen Shot 2022-03-31 at 11 01 29 PM

**Comment by yangwao** [comment]
seems to work well, will wait for code review 

image


image

**Review by yangwao — APPROVED** [review]

File: components/base/CarouselCardList.vue [hunk 1]
```diff
@@ -71,9 +71,10 @@
 
 <script lang="ts">
 import { Component, mixins, Prop } from 'nuxt-property-decorator'
-// import Identicon from '@polkadot/vue-identicon'
 import AuthMixin from '@/utils/mixins/authMixin'
 
+import type { CarouselNFTs } from './types'
```
Comment (vikiival) [inline components/base/CarouselCardList.vue:76]: 🙅‍♂️ never ever use `s`. With a combination with array I see in the next line is a good combo for the Bug :)

File: queries/rmrk/subsquid/collectionEntityById.graphql [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+      name
+      price
+      id
+      currentOwner
+      issuer
+      meta {
+        id
+        image
+      }
+      updatedAt
```
Comment (vikiival) [inline queries/rmrk/subsquid/collectionEntityById.graphql:19]: make a fragment please so we do not copy paste the variables

```graphql
fragment nft on NFTEntity {
  name
  price
  id
  currentOwner
  issuer
}

fragment baseMeta on Metadata {
  id
  image
}
```

File: components/rmrk/Gallery/VisitedNFT.vue [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+@Component({
+  components: {
+    CarouselCardList: () => import('@/components/base/CarouselCardList.vue'),
+  },
+})
+/**
+ * class name
+ */
+export default class VisitedNFT extends Vue {
+  MIN_NFTS = 3
```
Comment (vikiival) [inline components/rmrk/Gallery/VisitedNFT.vue:29]: move to constatnts

File: components/rmrk/Gallery/VisitedNFT.vue [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  nfts: CarouselNFTs[] = []
```
Comment (vikiival) [inline components/rmrk/Gallery/VisitedNFT.vue:30]: missing visibility modifier 👁️

**Review by vikiival — CHANGES_REQUESTED** [review]
Let's make the code more reusable 🥺

**Comment by vikiival** [comment]
Also both new components are doing the same logic :) it would be far better to make one common component which can tackle both.

File: components/rmrk/Gallery/VisitedNFT.vue (hunk 3 above)
Comment (preschian) [reply components/rmrk/Gallery/VisitedNFT.vue:29]: noted

File: components/rmrk/Gallery/VisitedNFT.vue (hunk 4 above)
Comment (preschian) [reply components/rmrk/Gallery/VisitedNFT.vue:30]: noted

File: queries/rmrk/subsquid/collectionEntityById.graphql (hunk 2 above)
Comment (preschian) [reply queries/rmrk/subsquid/collectionEntityById.graphql:19]: hi @vikiival, need your help with this
1. I tried to use this one https://github.com/kodadot/nft-gallery/blob/main/queries/fragments/nft.graphql got the error ```Cannot query field \"collectionId\" on type \"NFTEntity\". Did you mean \"collection\"?```
2. created `baseMeta` got this error `Unknown type \"Metadata\".`

did I miss something?
my updated query:
```graphql
#import "../../fragments/nft.graphql"
#import "../../fragments/nftDetails.graphql"
#import "../../fragments/baseMeta.graphql"

query collectionEntityById($id: ID!, $nftId: ID!) {
  collectionEntityById(id: $id) {
    id
    name
    nfts(
      limit: 10
      orderBy: [price_ASC, updatedAt_DESC]
      where: { price_gt: "0", id_not_eq: $nftId }
    ) {
      ...nftDetails
      meta {
        ...baseMeta
      }
      updatedAt
    }
  }
}

```

Comment (vikiival) [reply queries/rmrk/subsquid/collectionEntityById.graphql:19]: Yup my bad
```graphql
fragment baseMeta on MetadataEntity {
  id
  image
}
```

Comment (preschian) [reply queries/rmrk/subsquid/collectionEntityById.graphql]: hi @vikiival, how about the 1st one. should I refactor the others or leave at it is

Comment (vikiival) [reply queries/rmrk/subsquid/collectionEntityById.graphql]: > how about the 1st one 

I missed that. Let's create a new nftdetail for subsquid and we will refactor that later 💯

File: components/rmrk/Gallery/GalleryItemRelated.vue [hunk 5]
```diff
@@ -0,0 +1,106 @@
+<template>
+  <div>
+    <div class="my-5" v-if="nftRelated.length >= MIN_NFTS">
+      <p class="subtitle is-size-4">More From This Collection</p>
+      <CarouselCardList :nfts="nftRelated" />
+    </div>
+    <div class="my-5" v-if="nftVisited.length >= MIN_NFTS">
+      <p class="subtitle is-size-4">Visited NFTs</p>
+      <CarouselCardList :nfts="nftVisited" />
+    </div>
```
Comment (vikiival) [inline components/rmrk/Gallery/GalleryItemRelated.vue:10]: 🙈

**Review by vikiival — CHANGES_REQUESTED** [review]
🙈

File: components/rmrk/Gallery/GalleryItemRelated.vue (hunk 5 above)
Comment (vikiival) [reply components/rmrk/Gallery/GalleryItemRelated.vue:10]: This should be component
```vue
    <div class="my-5" v-if="nftRelated.length >= MIN_NFTS">
      <p class="subtitle is-size-4">More From This Collection</p>
      <CarouselCardList :nfts="nftRelated" />
    </div>
```

also missing translations

Comment (preschian) [reply components/rmrk/Gallery/GalleryItemRelated.vue:10]: @vikiival noted. do you have any concerns in the `async fetch()` section?

**Comment by preschian** [comment]
DeepScan scans unrelated files in this PR 🤔

Screen Shot 2022-04-01 at 6 07 39 PM

File: components/rmrk/Gallery/GalleryItemRelated.vue [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+   * Nuxt built-in data fetching
+   * https://nuxtjs.org/docs/features/data-fetching/
+   */
+  async fetch() {
+    // check visited NFT from localStorage.history
+    const getHistory = localStorage.getItem('history')
+    const {
+      history: { visitedNFTs },
+    } = getHistory && JSON.parse(getHistory)
+    const ids = visitedNFTs?.map((nft) => nft.id)
```
Comment (vikiival) [inline components/rmrk/Gallery/GalleryItemRelated.vue:52]: utiiiiils

File: components/rmrk/Gallery/GalleryItemRelated.vue [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+    return data.map((nft) => ({
+      ...nft,
+      timestamp: formatDistanceToNow(new Date(nft.updatedAt), {
+        addSuffix: true,
+      }),
+      image: imageOf(nft.meta.id, nft.meta.image),
+    }))
+  }
+}
```
Comment (vikiival) [inline components/rmrk/Gallery/GalleryItemRelated.vue:105]: utiiills

File: components/rmrk/Gallery/GalleryItemRelated.vue [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    this.nftVisited = visited || []
+  }
+
+  /**
+   * Format the data to fit with CarouselNFT[]
+   * Get cloudflare images
+   * Update timestamp
+   */
+  public async formatNFT(data) {
+    if (!data) return undefined
```
Comment (vikiival) [inline components/rmrk/Gallery/GalleryItemRelated.vue:92]: Mate pls don't do this

Comment (vikiival) [reply components/rmrk/Gallery/GalleryItemRelated.vue:92]:
```suggestion
    if (!data) {
      return undefined
    }
```

File: components/rmrk/Gallery/GalleryItemRelated.vue [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+
+    const [dataRelated, dataVisited] = await Promise.all(queries)
+    const [related, visited] = await Promise.all([
+      dataRelated &&
+        (await this.formatNFT(dataRelated.data?.collectionEntityById?.nfts)),
+      dataVisited && (await this.formatNFT(dataVisited.data?.nftEntities)),
+    ])
+
+    this.nftRelated = related || []
+    this.nftVisited = visited || []
```
Comment (vikiival) [inline components/rmrk/Gallery/GalleryItemRelated.vue:83]:
```suggestion
    this.nftRelated = related ?? []
    this.nftVisited = visited ?? []
```

File: components/rmrk/Gallery/GalleryItemRelated.vue [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+
+    const queries = [
+      this.collectionLength > this.MIN_NFTS && (await queryRelated),
+      ids?.length >= this.MIN_NFTS && (await queryVisited),
+    ]
+
+    const [dataRelated, dataVisited] = await Promise.all(queries)
+    const [related, visited] = await Promise.all([
+      dataRelated &&
+        (await this.formatNFT(dataRelated.data?.collectionEntityById?.nfts)),
```
Comment (vikiival) [inline components/rmrk/Gallery/GalleryItemRelated.vue:78]: why `await` in awaited `Promise.all` ?

**Review by vikiival — CHANGES_REQUESTED** [review]

File: components/rmrk/Gallery/GalleryItemRelated.vue (hunk 6 above)
Comment (preschian) [reply components/rmrk/Gallery/GalleryItemRelated.vue:52]: Hi @vikiival, can you elaborate more about this? What do utils mean?

File: components/rmrk/Gallery/GalleryItemRelated.vue (hunk 9 above)
Comment (preschian) [reply components/rmrk/Gallery/GalleryItemRelated.vue:83]: when I try to change this, got error in my IDE
```
Type 'false | CarouselNFT[]' is not assignable to type 'CarouselNFT[]'.
  Type 'boolean' is not assignable to type 'CarouselNFT[]'.Vetur(2322)
```

File: components/rmrk/Gallery/GalleryItemRelated.vue (hunk 10 above)
Comment (preschian) [reply components/rmrk/Gallery/GalleryItemRelated.vue:78]: need that for `this.formatNFT`

File: components/rmrk/Gallery/GalleryItemRelated.vue (hunk 6 above)
Comment (preschian) [reply components/rmrk/Gallery/GalleryItemRelated.vue:52]: oohh, did you mean to put this section into `utils/localstorage.ts` for example?

File: components/rmrk/Gallery/GalleryItemRelated.vue (hunk 7 above)
Comment (preschian) [reply components/rmrk/Gallery/GalleryItemRelated.vue]: updated, moved this section to utils

File: components/rmrk/Gallery/GalleryItemRelated.vue [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+    this.nftRelated = related || []
+    this.nftVisited = visited || []
+  }
+
+  /**
+   * Format the data to fit with CarouselNFT[]
+   * Get cloudflare images
+   * Update timestamp
+   */
+  public async formatNFT(data) {
```
[the remark this answers is not in the document]
Comment (preschian) [reply components/rmrk/Gallery/GalleryItemRelated.vue]: updated in the utils

File: components/rmrk/Gallery/GalleryItemRelated.vue (hunk 10 above)
Comment (preschian) [reply components/rmrk/Gallery/GalleryItemRelated.vue]: tried to remove `await` but got error, did I miss something?

File: components/rmrk/Gallery/GalleryItemRelated.vue (hunk 6 above)
Comment (yangwao) [reply components/rmrk/Gallery/GalleryItemRelated.vue]: yup, I guess so!

**Comment by preschian** [comment]
oh, it seems not triggering on route change (homepage to item), checking

**Comment by preschian** [comment]
updated, ready for review again 🙏🏻

File: components/rmrk/Gallery/GalleryItemCarousel.vue [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+    CarouselCardList: () => import('@/components/base/CarouselCardList.vue'),
+  },
+})
+/**
+ * class name
+ */
+export default class GalleryItemCarousel extends Vue {
+  @Prop({ type: String, required: true }) type!: 'related' | 'visited'
+  @Prop({ default: '' }) readonly collectionId!: string
+  @Provide() MIN_NFTS = 3
```
Comment (vikiival) [inline components/rmrk/Gallery/GalleryItemCarousel.vue:31]: `utils/constants.ts` cries in the corner

**Review by vikiival — CHANGES_REQUESTED** [review]
code looks definitely cleaner

just move the Min to the constants and make sure that **all** previous comments are resolved :)

File: components/rmrk/Gallery/GalleryItemCarousel.vue (hunk 12 above)
Comment (preschian) [reply components/rmrk/Gallery/GalleryItemCarousel.vue:31]: aahh, I see. will move that

Comment (preschian) [reply components/rmrk/Gallery/GalleryItemCarousel.vue]: updated

**Comment by preschian** [comment]
> code looks definitely cleaner
> 
> just move the Min to the constants and make sure that **all** previous comments are resolved :)

yaps, previous comments should be updated also. some of them outdated

**Review by vikiival — APPROVED** [review]
DID NOT test 🧪 

code is ~oki

**Comment by yangwao** [comment]
Found container 🫙 is broken on mobile again 😬 and it's moving, but I guess we can postpone it to another following issue.

https://user-images.githubusercontent.com/5887929/161374871-2edd68c8-ee2e-4395-833a-a5359da66682.mp4

**Comment by yangwao** [comment]
pay 300 usd

**Comment by yangwao** [comment]
😍 Perfect, I’ve sent the payout 
💵 $300 @ 195.63 USD/KSM ~ 1.534 $KSM
🧗 [DY4SQF2iD456tH89aQtz5wv1EV3BbSW8wKKuMcwbmXaj1pM](https://kusama.subscan.io/account/DY4SQF2iD456tH89aQtz5wv1EV3BbSW8wKKuMcwbmXaj1pM)
🔗 [0x22f7ec0072454afb40b92070fa0ed6f7c09aecb73511793a1c0e4d9db5151191](https://kusama.subscan.io/extrinsic/0x22f7ec0072454afb40b92070fa0ed6f7c09aecb73511793a1c0e4d9db5151191)

🪅 Let’s grab another issue and get rewarded! 
🪄 [github.com/kodadot/nft-gallery/issues](https://github.com/kodadot/nft-gallery/issues/)

**Comment by preschian** [comment]
> Found container 🫙 is broken on mobile again 😬 and it's moving, but I guess we can postpone it to another following issue.
> 
>  Record_2022-04-02-10-36-02.mp4

oops 😱

thanks a lot @yangwao @vikiival

**Pull request merged**